File: src/web/src/render/effect/RippleEffect.ts
import type { Effect } from './Effect.js';

export class RippleEffect implements Effect {
  static readonly DURATION_MS = 1500;
  static readonly RADIUS      = 75;
  static readonly WAVELENGTH  = 35;
  static readonly AMPLITUDE   = 10;
  static readonly SPEED       = 2.5;
  private static readonly TICK_MS = 16;

  private cx = 0;
  private cy = 0;
  private phase     = 0;
  private strength  = 0;
  private ticksLeft = 0;
  private timerId: ReturnType<typeof setInterval> | null = null;

  constructor(private readonly repaint: () => void) {}

  private totalTicks(): number {
    return RippleEffect.DURATION_MS / RippleEffect.TICK_MS;
  }

  trigger(x: number, y: number): void {
    this.cx        = x;
    this.cy        = y;
    this.phase     = 0;
    this.ticksLeft = this.totalTicks();
    this.strength  = 1;
    this.clearTimer();
    this.timerId = setInterval(() => this.tick(), RippleEffect.TICK_MS);
  }

  isActive(): boolean { return this.timerId !== null; }

  stop(): void { this.clearTimer(); }

  applyToPixels(src: Uint8ClampedArray, dst: Uint8ClampedArray, w: number, h: number): void {
    const { cx, cy, phase, strength } = this;
    const R  = RippleEffect.RADIUS;
    const WL = RippleEffect.WAVELENGTH;
    const A  = RippleEffect.AMPLITUDE;
    const x0 = Math.max(0, cx - R);
    const x1 = Math.min(w - 1, cx + R);
    const y0 = Math.max(0, cy - R);
    const y1 = Math.min(h - 1, cy + R);

    for (let oy = y0; oy <= y1; oy++) {
      for (let ox = x0; ox <= x1; ox++) {
        const ddx = ox - cx;
        const ddy = oy - cy;
        const d   = Math.sqrt(ddx * ddx + ddy * ddy);
        if (d === 0 || d >= R) continue;

        const t            = d / R;
        const falloff      = 1 - t;
        const displacement = A * strength * falloff
          * Math.sin(2 * Math.PI * (d / WL - phase));
        const sourceDist   = d - displacement;
        const dstBase      = (oy * w + ox) * 4;

        if (sourceDist <= 0) {
          const srcBase = dstBase;
          dst[dstBase]     = src[srcBase];
          dst[dstBase + 1] = src[srcBase + 1];
          dst[dstBase + 2] = src[srcBase + 2];
          dst[dstBase + 3] = src[srcBase + 3];
          continue;
        }

        const scale   = sourceDist / d;
        const ix      = Math.max(0, Math.min(w - 1, Math.round(cx + ddx * scale)));
        const iy      = Math.max(0, Math.min(h - 1, Math.round(cy + ddy * scale)));
        const srcBase = (iy * w + ix) * 4;

        dst[dstBase]     = src[srcBase];
        dst[dstBase + 1] = src[srcBase + 1];
        dst[dstBase + 2] = src[srcBase + 2];
        dst[dstBase + 3] = src[srcBase + 3];
      }
    }
  }

  private tick(): void {
    this.phase     += RippleEffect.SPEED * (RippleEffect.TICK_MS / 1000);
    this.ticksLeft -= 1;
    this.strength   = Math.max(0, this.ticksLeft / this.totalTicks());
    if (this.ticksLeft <= 0) {
      this.clearTimer();
    }
    this.repaint();
  }

  private clearTimer(): void {
    if (this.timerId !== null) {
      clearInterval(this.timerId);
      this.timerId = null;
    }
  }
}

File: src/main/java/com/ullgren/modern/simple3d/render/effect/RippleEffect.java
package com.ullgren.modern.simple3d.render.effect;

import javax.swing.Timer;

/**
 * Image-space ripple effect.
 * <p>
 * Concentric sine-wave rings radiate outward from the trigger point, fading over
 * {@link #DURATION_MS} milliseconds. Each pixel inside {@link #RADIUS} is displaced
 * radially by {@code AMPLITUDE × strength × falloff(d) × sin(2π × (d/WAVELENGTH - phase))},
 * where {@code phase} advances each tick to make rings appear to move outward.
 * <p>
 * A radial {@code falloff = 1 - d/RADIUS} smoothly fades the distortion at the edge.
 */
public class RippleEffect implements Effect {

  /** Duration of the full ripple animation. */
  public static final int   DURATION_MS = 1500;
  /** Outer boundary of the effect region in pixels. */
  public static final int   RADIUS      = 75;
  /** Distance in pixels between successive wave crests. */
  public static final float WAVELENGTH  = 35f;
  /** Peak radial displacement of a pixel in pixels. */
  public static final float AMPLITUDE   = 10f;
  /** Phase advance per second — how many full wave cycles pass the origin per second. */
  public static final float SPEED       = 2.5f;
  private static final int  TICK_MS     = 16;

  private int   cx, cy;
  private float phase;
  private float strength;
  private int   ticksLeft;

  private Timer timer;

  /**
   * @param repaint callback invoked each tick to request a canvas repaint
   */
  public RippleEffect(Runnable repaint) {
    timer = new Timer(TICK_MS, e -> {
      phase     += SPEED * (TICK_MS / 1000f);
      ticksLeft -= 1;
      strength   = Math.max(0f, (float) ticksLeft / totalTicks());
      if (ticksLeft <= 0) {
        timer.stop();
      }
      repaint.run();
    });
  }

  private int totalTicks() {
    return DURATION_MS / TICK_MS;
  }

  @Override
  public void trigger(int x, int y) {
    cx        = x;
    cy        = y;
    phase     = 0f;
    ticksLeft = totalTicks();
    strength  = 1f;
    timer.stop();
    timer.start();
  }

  @Override
  public boolean isActive() {
    return timer.isRunning();
  }

  @Override
  public void stop() {
    timer.stop();
  }

  @Override
  public void applyToPixels(int[] src, int[] dst, int w, int h) {
    int x0 = Math.max(0,     cx - RADIUS);
    int x1 = Math.min(w - 1, cx + RADIUS);
    int y0 = Math.max(0,     cy - RADIUS);
    int y1 = Math.min(h - 1, cy + RADIUS);

    for (int oy = y0; oy <= y1; oy++) {
      for (int ox = x0; ox <= x1; ox++) {
        float ddx = ox - cx;
        float ddy = oy - cy;
        float d   = (float) Math.sqrt(ddx * ddx + ddy * ddy);
        if (d == 0 || d >= RADIUS) continue;

        float t           = d / RADIUS;
        float falloff     = 1f - t;
        float displacement = AMPLITUDE * strength * falloff
            * (float) Math.sin(2 * Math.PI * (d / WAVELENGTH - phase));

        // Backward mapping: sourceDist = d - displacement.
        // Positive displacement pulls source from smaller radius → outward push.
        float sourceDist = d - displacement;
        if (sourceDist <= 0) {
          dst[oy * w + ox] = src[oy * w + ox];
          continue;
        }
        float scale = sourceDist / d;
        int   ix    = Math.max(0, Math.min(w - 1, Math.round(cx + ddx * scale)));
        int   iy    = Math.max(0, Math.min(h - 1, Math.round(cy + ddy * scale)));
        dst[oy * w + ox] = src[iy * w + ix];
      }
    }
  }
}
